v3/mqtt.go

```diff
@@ -465,91 +577,171 @@ func (conn *mqttConn) runPacketWriter(wg *sync.WaitGroup) {
 		wg.Done()
 	}()
 
-	// Run until closed
-	for p := range conn.outPktCh {
-		err := conn.writePacket(p)
-		if err != nil {
-			logError("Unable to write out packet %s", p.Type())
-			switch p.Type() {
-			case packet.PINGREQ:
-				conn.pingRespCh <- false
-			case packet.SUBSCRIBE:
-				subPkt := p.(*packet.SubscribePacket)
-				conn.queueAckCh <- MqttQueueResult{subPkt.PacketID, err}
-			case packet.PUBLISH:
-				pubPkt := p.(*packet.PublishPacket)
-				conn.queueAckCh <- MqttQueueResult{pubPkt.PacketID, err}
-			default:
-				// We don't have a packet ID (or did not expect to receive this
-				// packet type and fail in sending it
-				err := fmt.Errorf("Failed to queue packet %s", p.Type())
-				conn.queueAckCh <- MqttQueueResult{0, err}
+	shouldLoop := true
+	for shouldLoop {
+		// We read two channels for writing out packets. The directSend
+		// channel received synchronous, which has an unbuffered queue.
+		// The queuedSend can back up because the client already has an ID to
+		// check for the response.
+		// NOTE: Writes are sent on a bufio.Writer, so writes are almost
+		// guaranteed to succeed, even if they never reach the server. See the
+		// note in writePacket for more information.
+		select {
+		case pktSendData := <-conn.directSendPktCh:
+			resultCh := pktSendData.resultCh
+			// Verify that we have the write packet type for direct sends
+			if pktSendData.pkt == nil {
+				// Case where the channel is closed
+				shouldLoop = false
+				break
+			}
+			if conn.verifyPacketType(pktSendData.pkt.Type()) != directPacketSendType {
+				resultCh <- MqttResponse{
+					Err: errors.New("Packet type must be queued"),
+				}
+			} else {
+				resultCh <- MqttResponse{Err: conn.writePacket(pktSendData.pkt)}
+			}
+		case pktSendData := <-conn.queuedSendPktCh:
+			if pktSendData.pkt == nil {
+				// Case where the channel is closed
+				shouldLoop = false
+				break
+			}
+			pktID := uint16(0)
+			if pktSendData.pkt.Type() == packet.PUBLISH {
+				pubPkt := pktSendData.pkt.(*packet.PublishPacket)
+				pktID = pubPkt.PacketID
+			}
+			// Get the packet ID, if any, for errors
+			resultCh := pktSendData.resultCh
+			if conn.verifyPacketType(pktSendData.pkt.Type()) != queuedPacketSendType {
+				resultCh <- MqttResponse{
+					Err: errors.New("Packet type must be sent directly"),
+				}
+			} else {
```

Comment:
I think, this case is also that it can omit pattern of `else` statement .

Reply:
same as the other comment.